compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -458,6 +458,11 @@ pub enum PredicateKind<'tcx> {
     /// `const fn foobar<Foo: Bar>() {}`).
     Trait(TraitPredicate<'tcx>, Constness),
 
+    /// Equivalent to `Trait(SizedTraitPredicate, Constness::NotConst)` that type parameters and
+    /// associated types implicitly have. Used to differentiate them when giving trait bound
+    /// errors.
+    ImplicitSizedTrait(TraitPredicate<'tcx>),
```

Comment:
maybe just add a field to `TraitPredicate`? The additional variant requires a lot of noisy changes throughout the compiler.

Reply:
Adding a field to `TraitPredicate` caused fingerprinting issues that went away once I added a new variant.

Reply:
that's odd, do we have any manual eq/hash/stablehash impls on TraitPredicate?

Reply:
Now I remember what I did: I `impl Eq for TraitPredicate` so that `Sized` obligations would always eq regardless of provenance (which seeing that the only place that that is needed is in `compiler/rustc_trait_selection/src/traits/auto_trait.rs`, it might be fine to add a separate method for "agnostic equality". The problem I imagined was that in the future someone might make a change while not being aware of the issues that doesn't account for the existence of a separate method instead of just `==`.